chore: migration to unity 2020.3.0 f1 + SSAO support

BrianAmadori opened the pull request:
Further reading: https://forum.unity.com/threads/changes-to-the-webgl-loader-and-templates-introduced-in-unity-2020-1.817698/

Test link: https://play.decentraland.zone/branch/chore/unity-2020-3-0-f1/index.html?ENV=org

SSAO is disabled by default. Toggle it to ON by typing the following in your browser's console:

`unityInterface.SendMessageToUnity('Main', 'ToggleSSAO')`

Disable it again by typing the same command.

### Issues encountered (and will be nice to double check in the smoke testing):
- Kernel's loader had to be modified because we don't use the unity.json file anymore. Have to make sure the external rendering injection still works with remote 2020 artifacts.
- Transparencies stopped working, had to make changes to enable them again for all our shaders. We have to ensure the transparencies still work as intended.
- Bloom had to be re-configured, so it looks a bit different now, we have to check it looks OK for most of the scenes.
- Base resolution implementation was re-implemented, so we have to make sure is still working as intended

### Full changes overview:
- Kernel's loader `Instantiate` flow had to be modified. We no longer use `unity.json` for the loader configuration.
- CI had to be changed to target our new unity docker image
- Base resolution implementation is mostly removed because it doesn't work anymore, removed the kernel mouse event and canvas size hacks. Now we have to use `devicePixelRatio` to change base resolution.
- The main shaders used by the client had to be ported from scratch from the new URP version. (Lit and Toon Shader)
- Bloom values had to be re-adjusted to match (or maybe improve) the rendering style of the main branch.
- SSAO integrated for Lit and Toon shaders.
- Changed our Lit shader name from `DCL/LWRP/Lit` to `DCL/Universal Render Pipeline/Lit`.
- Had to regenerate our `shaderVariantsCollection` files because they were invalidated by the new shaders.

AjimenezDCL committed:
Project Ported

AjimenezDCL committed:
AO Compatibility with our shaders

AjimenezDCL committed:
Unity 2020.2b hacked into kernel

AjimenezDCL committed:
Now our DCL Lit shader works with SSAO in webgl

AjimenezDCL committed:
Remove WebGL postprocess to avoid the custom DCL template

AjimenezDCL committed:
fixing lint

AjimenezDCL committed:
fixing lint + disable unity from CI

AjimenezDCL committed:
disable unity from CI

AjimenezDCL committed:
disable unity from CI

AjimenezDCL committed:
$%@#$ lint

AjimenezDCL committed:
Avoid replacing the built-in unity build

AjimenezDCL committed:
Avoid replacing the built-in unity build

AjimenezDCL committed:
Fix some CI issues

AjimenezDCL committed:
More CI issues

AjimenezDCL committed:
More CI issues!

AjimenezDCL committed:
More CI issues!

AjimenezDCL committed:
Removed issue with mouse position

AjimenezDCL committed:
Tweaked SSAO

AjimenezDCL committed:
Tweaked SSAO

AjimenezDCL committed:
Fixing mouse mispositioning: attempt 1

AjimenezDCL committed:
Compiled Toon Shader for URP 10

AjimenezDCL committed:
Added DepthNormal pass and SSAO to toon shader

AjimenezDCL committed:
Tentative fix for avatar editor missing.

AjimenezDCL committed:
Updated to Unity 2020.2.0f1

Commit:
merge with master

Commit:
fix: resolution change not working on Unity 2020

Commit:
fix: pedantic linter fail

Commit:
chore: version update to 2020.3.0

Commit:
fix: index.html points to the proper .unityweb files + new build upload

Commit:
try to cache bust

Commit:
CI for 2020.3.0

Commit:
remove hardcoded build

Commit:
MainTest scene double event system issue

Commit:
MainTest lighting file

Commit:
SSAO tweaks

Commit:
backpack visual fix

Commit:
backpack visual fix (test 2)

Commit:
change library cache id

Commit:
revert test

Commit:
shader test

Commit:
update gitignore

Commit:
updated srp packages

Commit:
make custom SSAO copy for DCL bugfixes

Commit:
fix script name in ForwardRenderer

Commit:
meta fix

Commit:
test

Commit:
ssao patch 2

Commit:
ssao patch 3

Commit:
ssao patch 4

Commit:
test 5

Commit:
fix: displaced SSAO in avatar preview

- also disabled cameras when avatar editor is opened to improve performance

Commit:
SSAO tweaks

Commit:
proper porting of Lit and Unlit shaders

- matched implementation of latest URP version
- bloom tweaks to match old version the best as possible, had to hardcode a * 5 in the color emission value and change bloom values
- had to tweak SSAO values again because everything changed
- re-applied all our custom fixes for GLTF
- transparencies were fixed

Commit:
remove override SSAO shader because is not needed anymore

Commit:
update materials

Commit:
AO tweaks

Commit:
Remade toon shader integration w/SSAO

Commit:
transparency patch

Commit:
fix: disable SSAO and DepthNormals pass for transparent materials

- updated prebuilt shaders and variant collections

Commit:
transparency fix

Commit:
Merge branch 'master' into chore/unity-2020-3-0-f1

# Conflicts:
#	kernel/static/unity/Build/DCLUnityLoader.js
#	unity-client/Assets/ABConverter/Core.cs
#	unity-client/Assets/Scenes/Test/MainTest.unity
#	website/public/index.html

Commit:
kernel code cleanup

Commit:
Merge branch 'master' into chore/unity-2020-3-0-f1

# Conflicts:
#	website/scripts/hash_generator.js

Commit:
add SetSSAO method for debugging

Commit:
make SendMessageToUnity public to allow for advanced debugging

Commit:
fix shader issue

Commit:
culling controller test fix

Commit:
disable SSAO + replace SetSSAO with ToggleSSAO to avoid invalid payload issue (SendMessage doesn't accept lonely bools)

Commit:
code cleanup

Commit:
Merge branch 'master' into chore/unity-2020-3-0-f1

AjimenezDCL commented on unity-client/Assets/Scripts/MainScripts/DCL/Controllers/Camera/CameraController.cs:
```diff
@@ -60,6 +60,28 @@ private void Start()
+        else
+        {
+            camera.enabled = true;
+        }
```
```suggestion
        camera.enabled = !visibleState;
```

AjimenezDCL reviewed (approved):
Code-wise looks good (left a minor comment). Im approving but will test it regardless later.

sandrade-dcl reviewed (changes requested):
Code reviewed and it looks fine.

I have started to test the build and I will put here the things I go finding:

**Visual differences in the avatars:**
- Prod:
![image](https://user-images.githubusercontent.com/64659061/113021721-26f17100-9184-11eb-9040-81be08af29bf.png)

- Migration to 2020.3:
![image](https://user-images.githubusercontent.com/64659061/113021838-45f00300-9184-11eb-8a06-2fd1eb7cd3bd.png)

olavra commented:
Missing material for Alice in Tutorial

![image](https://user-images.githubusercontent.com/41490935/113030066-7e480f00-918d-11eb-98be-9ae2a8f1267c.png)

pravusjif reviewed (changes requested):
* The code looks great
* The build performed great (full-tested as always, and re-tested with SSAO ON)
* I'll add the visual issues I saw, you probably already know of them:
(all screenshots and vids where taken with the SSAO effect OFF)

1. There is some strange "mesh clipping" that apparently also happens in master's build but with another color:
- master build:
https://user-images.githubusercontent.com/1031741/113045466-4208b480-9175-11eb-85e0-95bd5d691208.mov

- this branch build:
https://user-images.githubusercontent.com/1031741/113045512-5482ee00-9175-11eb-9d89-bd5a6144bbcf.mov

- Seen in another user:
![Screen Shot 2021-03-30 at 16 11 59](https://user-images.githubusercontent.com/1031741/113045596-6b294500-9175-11eb-87da-6d6028ee97f6.png)

I believe it may be bound to the excessive rim lighting on the toon shader

2. Excessive rim lighting:
![Screen Shot 2021-03-30 at 16 10 57](https://user-images.githubusercontent.com/1031741/113045709-8bf19a80-9175-11eb-8489-b4b4cbbbd60c.png)

![no-SSAO-strong-rimlight](https://user-images.githubusercontent.com/1031741/113045832-ae83b380-9175-11eb-8bc4-9a709c691443.png)

3. Transparencies issue:
![Screen Shot 2021-03-30 at 16 07 58](https://user-images.githubusercontent.com/1031741/113045896-c0655680-9175-11eb-97a3-5ee2c8d3f376.png)
![no-SSAO-transparencies](https://user-images.githubusercontent.com/1031741/113045904-c2c7b080-9175-11eb-9ff8-8743a2174c1d.png)
![Screen Shot 2021-03-30 at 16 14 05](https://user-images.githubusercontent.com/1031741/113045907-c52a0a80-9175-11eb-91bb-80d36e5be4fe.png)
![Screen Shot 2021-03-30 at 16 00 30](https://user-images.githubusercontent.com/1031741/113045912-c5c2a100-9175-11eb-8331-e082ab0d840c.png)

Commit:
fix tutorial alice materials

Commit:
add alpha clip to shadervariants and prewarm assets

Commit:
update sprite atlases

Commit:
add VersionControlSettings file

Commit:
fix: washed up gif colors

Commit:
avatar AO + environment fresnel tweaks

Commit:
gif washed out fix 2

Commit:
revert fix

D4rWiNSS commented on unity-client/Logs/ApiUpdaterCheck.txt:
```diff
@@ -0,0 +1,10 @@
+C# parse time         : 335ms
+candidates check time : 51ms
+console write time    : 25ms
+
```
Should we include all the logs in the .gitignore?

D4rWiNSS reviewed (approved):
Good job! This will be a big thing!

Commit:
UI atlas update

Commit:
remove SSS properties from toon shader

Commit:
fix: error handler hook not working + localhost debugging not working

Commit:
disable rim lighting for toon shader

Commit:
add more subtlety to AO toon shader visuals

Commit:
apply PR feedback

Commit:
fix: kernel error handling issue

- the error handler override was throwing an error, had to extract a custom unminified preprocessed version of the loader from the Unity build tools source code. Took the opportunity to also extract the debug symbols support (albeit it seems to not fully work yet)

- HUDs were not initialized when running a local build using make watch. Turns out we weren't waiting for EngineStarted() properly. Fixed it.

- Some tweaks to .gitignore

- Tweaks to MakeFile and publish.sh to avoid having old Unity files in the NPM package

- Bump Configuration.cs version to 1.0

Commit:
re-apply big sur fix

Commit:
re-apply drawing buffer fix

sandrade-dcl reviewed (approved):
Tested and checked the visual fixes. Thanks Brian, good job!

menduz commented on unity-client/Logs/AssetImportWorker0-prev.log:
```diff
@@ -0,0 +1,182 @@
+Using pre-set license
```
Remove this file

menduz commented on unity-client/Logs/AssetImportWorker0.log:
```diff
@@ -0,0 +1,747 @@
+Using pre-set license
```
Remove this file

pravusjif reviewed (approved):
Tested everything again in ZONE 64,- 64 and in ORG and found no issues, great job!

menduz commented on kernel/packages/unity-interface/loader.ts:
```diff
@@ -39,8 +37,7 @@ async function injectRenderer(baseUrl: string): Promise<LoadRendererResult> {
     DclRenderer: globalThis.DclRenderer,
-    UnityLoader: globalThis.UnityLoader,
-    buildConfigPath
+    UnityLoader: globalThis.UnityLoader
```
```suggestion
    UnityLoader: globalThis.UnityLoader,
    baseUrl
```

menduz commented on kernel/packages/unity-interface/loader.ts:
```diff
@@ -8,7 +8,6 @@ export type UnityLoader = {}
 export type LoadRendererResult = {
   DclRenderer: DclRenderer
   UnityLoader: UnityLoader
```
```suggestion
  UnityLoader: UnityLoader
  baseUrl: string
```

menduz commented on kernel/packages/unity-interface/initializer.ts:
```diff
@@ -20,11 +20,11 @@ export type InitializeUnityResult = {
   const urlParams = new URLSearchParams(document.location.search)
 
-  const { buildConfigPath } = await loadUnity(urlParams.get('renderer') || undefined)
+  await loadUnity(urlParams.get('renderer') || undefined)
```
```suggestion
  const { baseUrl } = await loadUnity(urlParams.get('renderer') || undefined)
```

menduz commented on kernel/packages/unity-interface/initializer.ts:
```diff
@@ -20,11 +20,11 @@ export type InitializeUnityResult = {
   initShared()
 
-  globalThis.globalStore.dispatch(initializeRenderer(container, buildConfigPath))
+  globalThis.globalStore.dispatch(initializeRenderer(container))
```
```suggestion
  globalThis.globalStore.dispatch(initializeRenderer(container, baseUrl))
```

menduz commented on kernel/packages/unity-interface/loader.ts:
```diff
@@ -49,15 +46,13 @@ async function loadDefaultRenderer(): Promise<LoadRendererResult> {
       DclRenderer: globalThis.DclRenderer,
-      UnityLoader: globalThis.UnityLoader,
-      buildConfigPath
+      UnityLoader: globalThis.UnityLoader
```
```suggestion
      UnityLoader: globalThis.UnityLoader,
      baseUrl: getRendererArtifactsRoot()
```

menduz commented on website/public/index.html:
```diff
@@ -207,9 +207,50 @@
+              companyName: "Decentraland",
+              productName: "Decentraland World Client",
+              productVersion: "0.1",
+            };
```
Please move this into the initializeRenderer function, you must replace `%PUBLIC_URL%` by `baseUrl`

menduz commented on kernel/packages/shared/renderer/actions.ts:
```diff
@@ -5,8 +5,7 @@ import { UnityInterfaceContainer } from 'unity-interface/dcl'
 export const INITIALIZE_RENDERER = '[Request] Initialize renderer'
-export const initializeRenderer = (container: HTMLElement, buildConfigPath: string) =>
-  action(INITIALIZE_RENDERER, { container, buildConfigPath })
+export const initializeRenderer = (container: HTMLElement) => action(INITIALIZE_RENDERER, { container })
```
```suggestion
export const initializeRenderer = (container: HTMLElement, baseUrl: string) => action(INITIALIZE_RENDERER, { container, baseUrl })
```

menduz commented on kernel/packages/shared/renderer/sagas.ts:
```diff
@@ -52,20 +51,29 @@ export function* ensureRenderer() {
 
 function* initializeRenderer(action: InitializeRenderer) {
-  const { container, buildConfigPath } = action.payload
+  const { container } = action.payload
```
```suggestion
  const { container, baseUrl } = action.payload
```

Please move the JSON from the HTML to here, create the final URL using `baseUrl` instead of `%PUBLIC_URL%`

menduz reviewed (changes requested):
It looks good, although it will break the new load-remote-renderer flows and the preview. 

I recommend to first merge the new version of unity in the new renderer so we can test the whole integration before merging this PR.

Commit:
fix: enable passing of baseUrl to loader instead of having it hardcoded in index.html

Commit:
Merge branch 'master' into chore/unity-2020-3-0-f1

Commit:
preview mode fix

menduz committed:
Bucke up, and read the full commit message (...)

The original idea was to fix the preview with this new version of unity,
sadly this is what happened: https://www.youtube.com/watch?v=AbSehcT19u0

List of changes:

1. No more promises in the redux store, having a promise "instanceJS"
was an anti pattern. We now access both unity interface and browser
interface values without a promise.
2. (this may be the most important change) a custom DCLUnityLoader.js is
no longer necessary. We will use the one exposed by unity in the build.
This will lead us to more maintaiable code.
3. The saga that creates the "renderer" now is only one function instead
of three. Specific code for unity and WS modes was moved to other files.
4. MESSAGE_FROM_ENGINE is no longer a saga action. It was pure overhead
for no reason. Messages are now handled directly by the interfaces and
wired at initialization time. (initializeRenderer saga)
5. `gifProcessor` was moved from the DCL global object to the globalThis

Bottom line: the emited code by this newer Unity version is _way_
superior in good practices to the previous code. Global errors nolonger
break unity, and proper stack traces are shown. Not only that, but if an
exception happen, it nolonger freezes the renderer in an unrecoverable
state. We should keep the generated loader.js as pure as possible now
that it doesn't require hacks.

menduz committed:
Merge branch 'chore/unity-2020-3-0-f1' of github.com:decentraland/explorer into chore/unity-2020-3-0-f1

menduz committed:
update preview

menduz committed:
fix: linter & deps

menduz committed:
Merge branch 'master' into chore/unity-2020-3-0-f1

menduz committed:
fix: linter & deps

menduz committed:
Merge branch 'chore/unity-2020-3-0-f1' of github.com:decentraland/explorer into chore/unity-2020-3-0-f1

menduz commented:
@pravusjif I changed the GIF to live in the globalObject instead of DCL please check that it didn't break anything. Smoke tests were OK

menduz committed:
remove unused script

menduz committed:
Merge branch 'master' into chore/unity-2020-3-0-f1

pravusjif commented:
the last build performed better than ever for me, and didn't see any issue at all.

BrianAmadori commented on kernel/packages/shared/world/TeleportController.ts:
```diff
@@ -70,6 +70,7 @@ export const CAMPAIGN_PARCEL_SEQUENCE = [
   { x: -109, y: -89 }
 ]
 
+// TODO: don't do classess if it holds no state. Use namespaces or functions instead.
```
Those are my java-isms. Sorry!
